File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;

public class Main {

	static class Person {
		long height;
		long count;

		public Person(long height, long count) {
			this.height = height;
			this.count = count;
		}
	}

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

		int n = Integer.parseInt(br.readLine());
		ArrayDeque<Person> stack = new ArrayDeque<>();
		long result = 0;

		for (int i = 0; i < n; i++) {
			long height = Long.parseLong(br.readLine());
			long count = 1;

			while (!stack.isEmpty()) {
				if (stack.peek().height <= height) {
					result += stack.peek().count;
					if (stack.peek().height == height) {
						count += stack.peek().count;
					}
					stack.pop();
				} else {
					break;
				}
			}

			if (!stack.isEmpty()) {
				result++;
			}

			stack.push(new Person(height, count));
		}

		System.out.println(result);
	}
}

File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.py
import sys


class Person:
    def __init__(self, height, count):
        self.height = height
        self.count = count


n = int(sys.stdin.readline().rstrip())
stack = []
result = 0

for _ in range(n):
    height = int(sys.stdin.readline().rstrip())
    count = 1

    while stack:
        if stack[-1].height <= height:
            result += stack[-1].count
            if stack[-1].height == height:
                count += stack[-1].count
            stack.pop()
        else:
            break

    if stack:
        result += 1

    stack.append(Person(height, count))

print(result)
